node/network/collator-protocol/src/collator_side.rs

```diff
@@ -580,6 +594,7 @@ where
 			handle_peer_view_change(ctx, state, peer_id, view).await?;
 		}
 		PeerDisconnected(peer_id) => {
+			state.known_validators.retain(|_, v| *v != peer_id);
```

Comment:
what if a peer disconnects and then reconnects again?
do we want to keep the `known_validators` anyway?
and `peer_views.entry(peer_id).or_default()` on `PeerConnected`?
also not clear how to handle `Declare` messages in case of reconnects

Reply:
> what if a peer disconnects and then reconnects again?
> do we want to keep the known_validators anyway?

I don't think so. At least we have to free these records somehow otherwise we are walking into a potential DoS.

> and peer_views.entry(peer_id).or_default() on PeerConnected?

why, this is a NOP now because it's handled by the validator discovery API?

> also not clear how to handle Declare messages in case of reconnects

re-declare in that case I think as a handshake for the protocol. This info may be "outdated" and the either the validator or collator may "remember" the wrong thing here.

Reply:
But it seems to me if a peer disconnects and reconnects again, it will not be added back to `known_validators`, and therefore we will not be advertising to this peer anymore.
Yeah, I understand there is a concern about unbounded growth.

> why, this is a NOP now because it's handled by the validator discovery API?

I was thinking about reconnects again. But I guess in this case we will receive `PeerViewChange` on a new connection, so it's fine.